# 评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

 导入需要的包

In [68]:
import pandas as pd

In [69]:
import matplotlib.pyplot as plt

导入原始数据定义为orignal_data

In [70]:
orignal_data=pd.read_csv("e_commerce.csv")

In [71]:
orignal_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 数据评估

### 抽样观察（先抽取20份作为样本观察）

In [72]:
orignal_data.sample(20)

InvoiceNo StockCode                         Description  Quantity  \
124418    546977     21116                        OWL DOORSTOP         1   
72414    C542245     22380                 TOY TIDY SPACEBOY          -1   
508976    579286     23102      SILVER HEARTS TABLE DECORATION         1   
146301    548984    85049C             ROMANTIC PINKS RIBBONS          1   
522249    580393    72351A  SET/6 TURQUOISE BUTTERFLY T-LIGHTS        12   
7331      537034     22738           RIBBON REEL SNOWY VILLAGE        10   
386364    570232    84030E       ENGLISH ROSE HOT WATER BOTTLE         2   
331180    565994     37500         TEA TIME TEAPOT IN GIFT BOX         1   
485825    577689     22945      CHRISTMAS METAL TAGS ASSORTED         12   
290924    562420     22419                    LIPSTICK PEN RED        50   
300099    563193     22139    RETROSPOT TEA SET CERAMIC 11 PC          1   
278583    561217     21716     BOYS VINTAGE TIN SEASIDE BUCKET         2   
74602     542503     35653               VINTAGE BEAD NOTEBOOK         6   
345763    567179    47590B         PINK HAPPY BIRTHDAY BUNTING         2   
432551    573876     22231          JIGSAW TREE WITH BIRDHOUSE         5   
48461     540468     22342          HOME GARLAND PAINTED ZINC          2   
319129    564840     23352       ROLL WRAP 50'S RED CHRISTMAS          3   
410764    572129     82551              LAUNDRY 15C METAL SIGN        12   
247224    558760    85123A  WHITE HANGING HEART T-LIGHT HOLDER         6   
406931    571840     22952     60 CAKE CASES VINTAGE CHRISTMAS        24   

             InvoiceDate  UnitPrice  CustomerID         Country  
124418   3/18/2011 12:08       8.29         NaN  United Kingdom  
72414    1/26/2011 15:06       2.10     13098.0  United Kingdom  
508976  11/29/2011 10:57       0.83     17772.0  United Kingdom  
146301    4/5/2011 12:37       1.25     14546.0  United Kingdom  
522249   12/4/2011 10:51       2.10     16438.0  United Kingdom  
7331     12/3/2010 17:20       1.65     13081.0  United Kingdom  
386364   10/9/2011 13:58       4.25     14040.0  United Kingdom  
331180    9/8/2011 11:46       4.95     13244.0  United Kingdom  
485825  11/21/2011 11:29       0.85     17664.0  United Kingdom  
290924    8/4/2011 16:38       0.83         NaN  United Kingdom  
300099   8/12/2011 16:30      10.79         NaN  United Kingdom  
278583   7/25/2011 17:09       4.96         NaN  United Kingdom  
74602    1/28/2011 11:58       2.95     16327.0  United Kingdom  
345763   9/18/2011 14:27       5.45     15547.0  United Kingdom  
432551   11/1/2011 12:48       0.59     16360.0  United Kingdom  
48461     1/7/2011 13:55       1.66         NaN  United Kingdom  
319129   8/30/2011 12:49       2.46         NaN  United Kingdom  
410764   10/21/2011 9:25       1.45     17528.0  United Kingdom  
247224    7/3/2011 12:38       2.95     16918.0  United Kingdom  
406931  10/19/2011 12:16       0.55     13590.0  United Kingdom

In [73]:
orignal_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


#### 数据一共有541909行，其中Description和CustomerID有缺失。invoicedata和customer ID数据类型有问题。

### 评估缺失数据

了解Description缺失数据的具体情况

观察发现Description缺失的数据UnitPrice都为0，验证猜想增加条件看是否有数据在Description缺失时UnitPrice不为0.

In [101]:
orignal_data[(orignal_data["Description"].isnull())&(orignal_data["UnitPrice"]!=0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

结论：根据目标要求根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。根据检索结果发现Description缺失时UnitPrice同时为0。
其中Description为商品名称，UnitPrice为单价，都是重要的评估值，所以在根据目标分析时应该删除掉这些干扰信息。

重复操作对CustomerID有缺失的值进行分析

In [75]:
orignal_data[orignal_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

通过观察在CustomerID缺失的数据中其他数据基本正常，不影响之后分析，所以保留该数据。

### 评估重复数据

首先我知道InvoiceNo: 发票号码，StockCode产品代码和CustomerID用户编号都是唯一标识符。所以对其进行重复数据评估。
但通过数据本身特点，因为同一次交易可以买多件不同商品，所以InvoiceNo可以重复，StockCode产品代码和CustomerID在本次数据集中允许重复，因此无需对数据进行重复评估。

### 评估不一致数据

在对数据集数据充分了解下可以进行，特别在同一标签可以有不同叫法名称的情况，例如本集中的Country可以检查是否有多个指代同一个国家的情况。

下面提取统计Country的各种标签以及个数

In [76]:
orignal_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


通过观察发现United Kingdom，UK，U.K.这三个标签都是表示的英国，United States和USA都表示的是美国存在重复。所以需要对数据进行处理。

### 对这两种类别的数据进行统合

## 评估无效和错误数据

In [77]:
orignal_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

观察发现Quantity数量和Unitprice单价存在负数，和实际意义不符合，首先筛选出这些数据分析其原因

In [78]:
orignal_data[orignal_data["Quantity"]<0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

分析发现Quantity为负数时`InvoiceNo`: 发票号码前面都是以C开头，表示订单被取消，所以进一步验证结果

In [79]:
orignal_data[(orignal_data["Quantity"]<0)  & (orignal_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

发现其中还有很多InvoiceNo开头不为C的数据，继续观察发现其Unitprice都为0，那么继续添加条件进行检索。

In [80]:
orignal_data[(orignal_data["Quantity"]<0)  & (orignal_data["InvoiceNo"].str[0] != "C") & (orignal_data["UnitPrice"]!=0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

发现三个检索条件后就没有检索值，说明猜想正确。然后更具数据分析的目标，对于InvoiceNo开头为C的订单是被取消的订单和UnitPrice为0的订单对于我们数据处理没有帮助，所以可以把这类数据删除。

In [81]:
orignal_data[orignal_data["UnitPrice"]<0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

发现UnitPrice小于0 的数据是出现了坏帐，所以也需要对该类数据进行处理。

## 数据清理

总结前面的数据过程，我们需要处理的数据包括
1.数据类型的转换
InvoiceDate 的数据数据类型转化为时间，CustomerID的数据类型转换成字符串
2.评估缺失数据
删除Description缺失的数据
3.重复数据评估
无
4.评估不一致数据
将标签代表类型相同的数据进行合并，United Kingdom，UK，U.K.这三个标签都是表示的英国可以合并，United States和USA都表示的是美国可以合并。
即把UK，U.K.替换为United Kingdom，把USA替换为United States。
5.评估无效和错误数据
发现Quantity数量和Unitprice单价存在负数，和实际意义不符合,经评估后发现可以删除

为了区分原始数据和清洗后的数据，创建新的变量cleaned_data来储存我们处理后的数据

In [82]:
cleaned_data = orignal_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

#### 1.InvoiceDate 的数据数据类型转化为时间，CustomerID的数据类型转换成字符串

In [83]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])

In [84]:
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [85]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)

In [86]:
cleaned_data["CustomerID"]=cleaned_data["CustomerID"].str.slice(0,5)

In [87]:
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom

经检查数据类型已经修改完成

#### 2.删除有空缺值的无效数据

In [88]:
cleaned_data.dropna(subset=['Description'], inplace=True)

In [89]:
cleaned_data[cleaned_data["Description"].isnull()]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

经检索已经删除完成

#### 4.评估不一致数据

将标签代表类型相同的数据进行合并，United Kingdom，UK，U.K.这三个标签都是表示的英国可以合并，United States和USA都表示的是美国可以合并。
即把UK，U.K.替换为United Kingdom，把USA替换为United States。

In [90]:
cleaned_data.replace("UK","United Kingdom", inplace=True)

In [91]:
cleaned_data.replace("U.K.","United Kingdom", inplace=True)

In [92]:
cleaned_data.replace("USA","United States", inplace=True)

In [93]:
cleaned_data["Country"].value_counts()

Country
United Kingdom          494024
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
United States              291
China                      288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


经检查标签名称以全部替换没有相同意义的标签

#### 5.评估无效和错误数据

发现Quantity数量和Unitprice单价存在负数，和实际意义不符合,经评估后发现可以删除

思路是把所以Quantity大于0的数据筛选出来重新附给Quantity。

In [94]:
cleaned_data = cleaned_data[cleaned_data["Quantity"]>=0]

In [95]:
cleaned_data[cleaned_data["Quantity"]<0]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

In [96]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"]>=0]
cleaned_data[cleaned_data["Quantity"]<0]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

经检查已经去除掉了所以的无效和错误数据

## 最后经过检查清洗后的数据集展示

In [97]:
cleaned_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-12-09 12:50:00       0.85      12680          France  
541905 2011-12-09 12:50:00       2.10      12680          France  
541906 2011-12-09 12:50:00       4.15      12680          France  
541907 2011-12-09 12:50:00       4.15      12680          France  
541908 2011-12-09 12:50:00       4.95      12680          France  

[530691 rows x 8 columns]

## 最后再随机取样30个观察数据是否符合我们的需要

In [98]:
cleaned_data.sample(30)

InvoiceNo StockCode                          Description  Quantity  \
333088    566171     20725              LUNCH BAG RED RETROSPOT         2   
227415    556876    47593B                SCOTTIE DOGS BABY BIB         1   
501635    578827     23434     3 RAFFIA RIBBONS 50'S CHRISTMAS          4   
65551     541696     21494       ROTATING LEAVES T-LIGHT HOLDER         7   
65871     541696    90202D          PINK ENAMEL FLOWER HAIR TIE         2   
279599    561332     22690        DOORMAT HOME SWEET HOME BLUE          2   
521091    580311     22776        SWEETHEART 3 TIER CAKE STAND          3   
279136    561241     84380    SET OF 3 BUTTERFLY COOKIE CUTTERS        12   
452763    575384     22941         CHRISTMAS LIGHTS 10 REINDEER         4   
5274      536856     22748            POPPY'S PLAYHOUSE KITCHEN         6   
450997    575177     23312          VINTAGE CHRISTMAS GIFT SACK         1   
69852     541987     22480         RED TEA TOWEL CLASSIC DESIGN         3   
67841     541847     21171                 BATHROOM METAL SIGN          2   
420399    572893    35819B           ACRYLIC HANGING JEWEL,BLUE        24   
517131    580032     21328               BALLOONS  WRITING SET          3   
185608    552822    84912B                   GREEN ROSE WASHBAG         1   
427713    573396     22065       CHRISTMAS PUDDING TRINKET POT          4   
434598    574053     22077               6 RIBBONS RUSTIC CHARM        12   
429148    573545     22947           WOODEN ADVENT CALENDAR RED         6   
128549    547357     22858                  EASTER TIN KEEPSAKE         6   
385005    570187     22424               ENAMEL BREAD BIN CREAM         1   
142424    548612     22851   SET 20 NAPKINS FAIRY CAKES DESIGN          1   
220090    556116    82616C  MIDNIGHT GLAMOUR SCARF KNITTING KIT        12   
155277    549981     22555            PLASTERS IN TIN STRONGMAN         1   
163910    550630     21394                 RED POLKADOT BEAKER         48   
31610     538999     22798     ANTIQUE GLASS DRESSING TABLE POT         8   
66699     541811     21035      SET/2 RED RETROSPOT TEA TOWELS          1   
413263    572309     21993           FLORAL FOLK STATIONERY SET         4   
361517    568346     23292               SPACEBOY CHILDRENS CUP        30   
183080    552642     23171             REGENCY TEA PLATE GREEN         12   

               InvoiceDate  UnitPrice CustomerID         Country  
333088 2011-09-09 12:22:00       1.65      18118  United Kingdom  
227415 2011-06-15 11:56:00       0.39      16241  United Kingdom  
501635 2011-11-25 14:25:00       0.83        nan  United Kingdom  
65551  2011-01-20 18:08:00       1.63        nan  United Kingdom  
65871  2011-01-20 18:08:00       2.90        nan  United Kingdom  
279599 2011-07-26 14:25:00       7.95      13492     Switzerland  
521091 2011-12-02 14:48:00       9.95      18245  United Kingdom  
279136 2011-07-26 11:00:00       1.25      13930  United Kingdom  
452763 2011-11-09 15:17:00       8.50      17690  United Kingdom  
5274   2010-12-03 10:26:00       2.10      13370  United Kingdom  
450997 2011-11-08 18:41:00       8.29        nan  United Kingdom  
69852  2011-01-24 16:05:00       1.25      14130  United Kingdom  
67841  2011-01-23 13:25:00       1.45      15443  United Kingdom  
420399 2011-10-26 14:36:00       0.38      18059  United Kingdom  
517131 2011-12-01 12:17:00       1.65      14659  United Kingdom  
185608 2011-05-11 12:59:00       3.75      15521  United Kingdom  
427713 2011-10-30 15:13:00       0.39      14723  United Kingdom  
434598 2011-11-02 14:04:00       1.95      16379  United Kingdom  
429148 2011-10-31 13:12:00       7.95      16969  United Kingdom  
128549 2011-03-22 12:15:00       1.65      17613  United Kingdom  
385005 2011-10-07 15:09:00      24.96        nan  United Kingdom  
142424 2011-04-01 11:34:00       0.85      13137  United Kingdom  
220090 2011-06-09 09:44:00       1.25      15214  United Kingdom  
155277 2011

## 保存清洗后的数据

In [99]:
cleaned_data.to_csv("e_commerce_cleaned.csv", index=False)
df=pd.read_csv("e_commerce_cleaned.csv")

C:\Users\CAshely\AppData\Local\Temp\ipykernel_16768\1074412508.py:2: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv("e_commerce_cleaned.csv")


In [100]:
df[df["Quantity"]<0]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []